File: lua/glean/diff.lua
-- glean.diff: parse unified-diff text into FileEntries / Hunks / DiffLines.
--
-- The parser is pure (no git, no IO) so it is trivially testable: feed it the
-- text produced by `git diff` and it returns an ordered list of FileEntries.
--
-- Each DiffLine carries the new-file line number (`new_lnum`) which is the
-- stable addressing basis for review marks: a `(commit, path, new_lnum)` tuple
-- never moves because a commit's post-image blob is immutable.
local M = {}

-- A DiffLine: { kind = "context"|"add"|"del", text, old_lnum, new_lnum }.
--   - add lines have a new_lnum but no old_lnum.
--   - del lines have an old_lnum but no new_lnum.
--   - context lines have both.

-- Parse a single `@@ -a,b +c,d @@` hunk header. The counts are optional and
-- default to 1 when omitted (git omits a count of 1).
local function parse_hunk_header(line)
  local old_start, old_count, new_start, new_count =
    line:match("^@@ %-(%d+),?(%d*) %+(%d+),?(%d*) @@")
  if not old_start then return nil end
  return {
    old_start = tonumber(old_start),
    old_count = tonumber(old_count) or 1,
    new_start = tonumber(new_start),
    new_count = tonumber(new_count) or 1,
    header = line,
    lines = {},
  }
end

-- Strip the `a/` or `b/` prefix git puts on diff paths, honoring quoted paths
-- only minimally (we keep it simple; review-sized diffs use plain paths).
local function strip_prefix(p)
  return (p:gsub("^[ab]/", ""))
end

-- Parse the path from a `diff --git a/<old> b/<new>` header. Returns
-- old_path, new_path (prefix-stripped). Falls back to nil on no match.
local function parse_git_header(line)
  local old_p, new_p = line:match("^diff %-%-git (%S+) (%S+)$")
  if not old_p then return nil, nil end
  return strip_prefix(old_p), strip_prefix(new_p)
end

-- Parse full unified-diff text (possibly spanning many files) into an ordered
-- list of FileEntries:
--   { path, old_path, kind = "add"|"delete"|"modify"|"rename", hunks }
-- `collapsed` is intentionally not set here; it is ephemeral view-state owned by
-- the renderer.
function M.parse(text)
  local files = {}
  local file = nil
  local hunk = nil
  local old_lnum, new_lnum = 0, 0

  local function finish_hunk()
    if file and hunk then
      file.hunks[#file.hunks + 1] = hunk
      hunk = nil
    end
  end

  local function finish_file()
    finish_hunk()
    file = nil
  end

  for line in (text .. "\n"):gmatch("(.-)\n") do
    local old_p, new_p = parse_git_header(line)
    if old_p then
      finish_file()
      file = {
        path = new_p,
        old_path = old_p,
        kind = "modify",
        hunks = {},
      }
      files[#files + 1] = file
    elseif file and not hunk and line:match("^new file mode") then
      file.kind = "add"
    elseif file and not hunk and line:match("^deleted file mode") then
      file.kind = "delete"
    elseif file and not hunk and line:match("^rename from ") then
      file.kind = "rename"
    elseif file and not hunk and line:match("^rename to ") then
      file.kind = "rename"
    elseif file and line:match("^%-%-%- ") then
      -- old-file marker; for added files this is `/dev/null`.
    elseif file and line:match("^%+%+%+ ") then
      local p = line:match("^%+%+%+ (.+)$")
      if p and p ~= "/dev/null" then file.path = strip_prefix(p) end
    elseif file and line:match("^@@") then
      finish_hunk()
      hunk = parse_hunk_header(line)
      if hunk then
        old_lnum = hunk.old_start
        new_lnum = hunk.new_start
      end
    elseif hunk then
      local marker = line:sub(1, 1)
      local body = line:sub(2)
      if marker == "+" then
        hunk.lines[#hunk.lines + 1] =
          { kind = "add", text = body, new_lnum = new_lnum }
        new_lnum = new_lnum + 1
      elseif marker == "-" then
        hunk.lines[#hunk.lines + 1] =
          { kind = "del", text = body, old_lnum = old_lnum }
        old_lnum = old_lnum + 1
      elseif marker == " " then
        hunk.lines[#hunk.lines + 1] =
          { kind = "context", text = body, old_lnum = old_lnum, new_lnum = new_lnum }
        old_lnum = old_lnum + 1
        new_lnum = new_lnum + 1
      elseif marker == "\\" then
        -- "\ No newline at end of file" — decoration, not a line.
      end
    end
  end

  finish_file()
  return files
end

return M

File: node/core/diff.ts
/**
 * Pure unified-diff parser: `git diff` text into FileEntries / Hunks / DiffLines.
 *
 * Every DiffLine carries a post-image slot (`newLnum`). A del's slot is the
 * post-image line it sat immediately before, so comment anchoring in the
 * post-image space is total.
 */
import type { RepoPath } from "./types.ts";

export type DiffLine =
  | { kind: "context"; text: string; oldLnum: number; newLnum: number }
  | { kind: "add"; text: string; newLnum: number }
  | { kind: "del"; text: string; oldLnum: number; newLnum: number };

export type Hunk = {
  oldStart: number;
  oldCount: number;
  newStart: number;
  newCount: number;
  header: string;
  lines: DiffLine[];
};

export type FileKind = "add" | "delete" | "modify" | "rename";

export type FileEntry = {
  path: RepoPath;
  oldPath: RepoPath;
  kind: FileKind;
  hunks: Hunk[];
};

const HUNK_HEADER = /^@@ -(\d+),?(\d*) \+(\d+),?(\d*) @@/;

function parseHunkHeader(line: string): Hunk | undefined {
  const m = HUNK_HEADER.exec(line);
  if (!m) return undefined;
  // git omits a count of 1.
  return {
    oldStart: Number(m[1]),
    oldCount: m[2] === "" ? 1 : Number(m[2]),
    newStart: Number(m[3]),
    newCount: m[4] === "" ? 1 : Number(m[4]),
    header: line,
    lines: [],
  };
}

function stripPrefix(p: string): RepoPath {
  return p.replace(/^[ab]\//, "") as RepoPath;
}

export function parse(text: string): FileEntry[] {
  const files: FileEntry[] = [];
  let file: FileEntry | undefined;
  let hunk: Hunk | undefined;
  let oldLnum = 0;
  let newLnum = 0;

  const finishHunk = () => {
    if (file && hunk) file.hunks.push(hunk);
    hunk = undefined;
  };

  for (const line of text.split("\n")) {
    const git = /^diff --git (\S+) (\S+)$/.exec(line);
    if (git) {
      finishHunk();
      file = {
        path: stripPrefix(git[2]!),
        oldPath: stripPrefix(git[1]!),
        kind: "modify",
        hunks: [],
      };
      files.push(file);
      continue;
    }
    if (!file) continue;
    if (!hunk) {
      if (line.startsWith("new file mode")) {
        file.kind = "add";
        continue;
      }
      if (line.startsWith("deleted file mode")) {
        file.kind = "delete";
        continue;
      }
      if (line.startsWith("rename from ") || line.startsWith("rename to ")) {
        file.kind = "rename";
        continue;
      }
      // Path markers are only valid before the first `@@`: inside a hunk
      // `--- x` is a deleted `-- x` line and `+++ x` an added `++ x` line.
      if (line.startsWith("--- ")) continue;
      if (line.startsWith("+++ ")) {
        const p = line.slice(4);
        if (p !== "" && p !== "/dev/null") file.path = stripPrefix(p);
        continue;
      }
    }
    if (line.startsWith("@@")) {
      finishHunk();
      hunk = parseHunkHeader(line);
      if (hunk) {
        oldLnum = hunk.oldStart;
        newLnum = hunk.newStart;
      }
      continue;
    }
    if (!hunk) continue;
    const marker = line.charAt(0);
    const body = line.slice(1);
    if (marker === "+") {
      hunk.lines.push({ kind: "add", text: body, newLnum });
      newLnum++;
    } else if (marker === "-") {
      hunk.lines.push({ kind: "del", text: body, oldLnum, newLnum });
      oldLnum++;
    } else if (marker === " ") {
      hunk.lines.push({ kind: "context", text: body, oldLnum, newLnum });
      oldLnum++;
      newLnum++;
    }
    // `\ No newline at end of file` is decoration, not a line.
  }
  finishHunk();
  return files;
}

/**
 * Map a pre-image line number through `hunks` to the post-image. Undefined when
 * the line did not survive (deleted or replaced).
 */
export function mapLnum(
  hunks: readonly Hunk[],
  lnum: number,
): number | undefined {
  let shift = 0;
  for (const h of hunks) {
    if (lnum < h.oldStart) break;
    if (lnum < h.oldStart + h.oldCount) {
      for (const dl of h.lines) {
        if (dl.kind !== "add" && dl.oldLnum === lnum) {
          return dl.kind === "context" ? dl.newLnum : undefined;
        }
      }
      return undefined;
    }
    shift = h.newStart + h.newCount - (h.oldStart + h.oldCount);
  }
  return lnum + shift;
}
